src/expire.c

```diff
@@ -161,15 +161,25 @@ void activeExpireCycle(int type) {
             unsigned long num, slots;
             long long now, ttl_sum;
             int ttl_samples;
-            iteration++;
 
             /* If there is nothing to expire try next DB ASAP. */
             if ((num = dictSize(db->expires)) == 0) {
                 db->avg_ttl = 0;
                 break;
             }
             slots = dictSlots(db->expires);
-            now = mstime();
+            now = ustime();
+
+            /* We can't block forever here even if there are many keys to
+             * expire. So after a given amount of milliseconds return to the
+             * caller waiting for the other active expire cycle. */
+            if (now > start + timelimit) {
+                timelimit_exit = 1;
+                server.stat_expired_time_cap_reached_count++;
+                break;
+            }
+
+            now /= 1000;//millisecond
```

Comment:
This makes sense because there's the `mstime()` syscall in iteration anyway so there's no point in adding the extra syscall and then reducing it to once every 16 iterations.
But I would suggest moving the timelimit check to **after** the expiration loop like it was before.